cajita/unit_test/tstSparseArray.hpp

```diff
@@ -0,0 +1,622 @@
+                         par_view( id, 1 ) - low_corner[1],
+                         par_view( id, 2 ) - low_corner[2] };
+            int grid_base[3] = {
+                static_cast<int>( std::lround( pos[0] * dx_inv ) - 1 ),
+                static_cast<int>( std::lround( pos[1] * dx_inv ) - 1 ),
+                static_cast<int>( std::lround( pos[2] * dx_inv ) - 1 ) };
+            int offset = 0;
+            for ( int i = 0; i <= 2; ++i )
+                for ( int j = 0; j <= 2; ++j )
+                    for ( int k = 0; k <= 2; ++k )
+                    {
+                        int cell_id[3] = { grid_base[0] + i, grid_base[1] + j,
+                                           grid_base[2] + k };
+                        auto tid = sparse_map.queryTile( cell_id[0], cell_id[1],
+                                                         cell_id[2] );
+                        qtid_res( id * 27 + offset ) = tid;
+                        offset++;
+                    }
+        } );
+    // check if all required cell are registered
+    auto qtid_mirror =
+        Kokkos::create_mirror_view_and_copy( Kokkos::HostSpace(), qtid_res );
+    for ( int i = 0; i < cell_num; ++i )
+    {
+        EXPECT_EQ( qtid_mirror( i ) < (int)test_array.numSoA(), true );
+        EXPECT_EQ( qtid_mirror( i ) >= 0, true );
+    }
+
+    // assign value
+    auto& map = test_layout.sparseMap();
+    // tile ijk, cell ijk, tile key
+    Kokkos::View<int* [7], TEST_DEVICE> info(
+        Kokkos::ViewAllocateWithoutInitializing( "tile_cell_info" ),
+        test_array.size() );
+    Kokkos::parallel_for(
+        "assign_value_to_sparse_cells",
+        Kokkos::RangePolicy<TEST_EXECSPACE>( 0, map.capacity() ),
+        KOKKOS_LAMBDA( const int index ) {
+            if ( map.valid_at( index ) )
+            {
+                auto tid = map.value_at( index );
+                auto tkey = map.key_at( index );
+                int ti, tj, tk;
+                map.key2ijk( tkey, ti, tj, tk );
+
+                for ( int ci = 0; ci < cell_per_tile_dim; ci++ )
+                    for ( int cj = 0; cj < cell_per_tile_dim; cj++ )
+                        for ( int ck = 0; ck < cell_per_tile_dim; ck++ )
+                        {
+                            // indices
+                            int cid = map.cell_local_id( ci, cj, ck );
+                            Kokkos::Array<int, 3> cell_ijk(
+                                { ti * cell_per_tile_dim + ci,
+                                  tj * cell_per_tile_dim + cj,
+                                  tk * cell_per_tile_dim + ck } );
+                            Kokkos::Array<int, 3> tile_ijk( { ti, tj, tk } );
+                            Kokkos::Array<int, 3> local_cell_ijk(
+                                { ci, cj, ck } );
+
+                            // access: cell ijk (- channel id)
+                            test_array.template get<0>( cell_ijk, 0 ) = ti;
```

Comment:
We should be able to reuse or imitate the `Cabana::SoA` cast functions to make this a little easier for the user

Reply:
Let's leave for now and plan to hide inside the parallel functions

Reply:
Sounds great